contracts/p1/RToken.sol

```diff
@@ -378,69 +337,72 @@ contract RTokenP1 is ComponentP1, ERC20PermitUpgradeable, IRToken {
+        _scaleUp(address(backingManager), baskets, totalSupply());
     }
 
     /// Melt a quantity of RToken from the caller's account, increasing the basket rate
     /// @param amtRToken {qRTok} The amtRToken to be melted
     /// @custom:protected
-    // checks: not trading paused or frozen
+    // checks: caller is furnace
     // effects:
     //   bal'[caller] = bal[caller] - amtRToken
     //   totalSupply' = totalSupply - amtRToken
-    //
-    // untestable:
-    //   `else` branch of `exchangeRateIsValidAfter` (ie. revert) shows as uncovered
-    //   but it is fully covered for `melt` (limitations of coverage plugin)
-    function melt(uint256 amtRToken) external exchangeRateIsValidAfter {
+    // BU exchange rate cannot decrease
+    // BU exchange rate CAN increase, but we already trust furnace to do this slowly
+    function melt(uint256 amtRToken) external {
         require(_msgSender() == address(furnace), "furnace only");
         _burn(_msgSender(), amtRToken);
         emit Melted(amtRToken);
     }
 
-    /// Dissolve an amount of RToken from caller's account and scale basketsNeeded down
+    /// Burn an amount of RToken from caller's account and scale basketsNeeded down
     /// Callable only by backingManager
     /// @param amount {qRTok}
     /// @custom:protected
-    function dissolve(uint256 amount) external notTradingPausedOrFrozen exchangeRateIsValidAfter {
+    // checks: caller is backingManager
+    // effects:
+    //   bal'[recipient] = bal[recipient] - amtRToken
+    //   totalSupply' = totalSupply - amtRToken
+    //   basketsNeeded' = basketsNeeded - baskets
+    // BU exchange rate cannot decrease, and it can only increase when < FIX_ONE.
+    function dissolve(uint256 amount) external {
         require(_msgSender() == address(backingManager), "not backing manager");
-        _redeem(_msgSender(), amount);
+        _scaleDown(_msgSender(), amount);
     }
 
     /// An affordance of last resort for Main in order to ensure re-capitalization
     /// @custom:protected
-    // checks: trading unpaused; unfrozen; caller is backingManager
+    // checks: caller is backingManager
     // effects: basketsNeeded' = basketsNeeded_
-    //
-    // untestable:
-    //   `else` branch of `exchangeRateIsValidAfter` (ie. revert) shows as uncovered
-    //   but it is fully covered for `setBasketsNeeded` (limitations of coverage plugin)
-    function setBasketsNeeded(uint192 basketsNeeded_)
-        external
-        notTradingPausedOrFrozen
-        exchangeRateIsValidAfter
-    {
+    function setBasketsNeeded(uint192 basketsNeeded_) external notTradingPausedOrFrozen {
         require(_msgSender() == address(backingManager), "not backing manager");
         emit BasketsNeededChanged(basketsNeeded, basketsNeeded_);
         basketsNeeded = basketsNeeded_;
+
+        // == P0 exchangeRateIsValidAfter modifier ==
+        uint256 supply = totalSupply();
+        require(supply > 0, "0 supply");
```

Comment:
this seems weird at first glance, but i think its ok?  there should not be a situation where bm is calling setBasketsNeeded when there is 0 supply, as it only does so during a haircut.

Reply:
Previously we were returning in this case and not doing the following checks. That seemed wrong: to change `basketsNeeded` without doing any checks